feat: add new functionality required by 7-sprint

AleksMosYanPrac opened the pull request:
feat: add new functionality required by 7-sprint
"possibility read/write data from/to CSV file"

test: add tests for new functionality

refactor: restructure project
refactor: add folder with resources to project
refactor: add example file with CSV data

fix: modify manager algorithms
"didn't update repositories after modifying Epic and Subtask"

AleksMosYanPrac committed:
feat: add new functionality required by 7-sprint
"possibility read/write data from/to CSV file"

test: add tests for new functionality

refactor: restructure project
refactor: add fileBackedTaskManager
style: clean empty line
chore: add DataSet.java to repository

refactor: add folder with resources to project
refactor: add example file with CSV data
refactor: add out folder to ignore

fix: modify manager algorithms
"didn't update repositories"

EugeneLenkevich commented on test/ru/yandex/practicum/kanban/repository/impls/in_file/InFileEpicRepositoryWithCSVFileDataSourceTest.java:
```diff
@@ -0,0 +1,155 @@
+            pathToFile = tempDir.resolve("temp_file_1.csv");
+            Files.createFile(pathToFile);
+            this.epicRepository = new InFileEpicRepositoryImpl(new CSVFileDataSource(pathToFile));
+        } catch (InvalidPathException ipe) {
```
В тесте можно не перехватывать исключение, junit и так распечатает стректейс ошибки :)

EugeneLenkevich commented on test/ru/yandex/practicum/kanban/repository/impls/in_file/datasource/CSVFileDataSourceTest.java:
```diff
@@ -0,0 +1,144 @@
+        dataSource.clear(typeQuery);
+        List<String> linesAfterClear = Files.readAllLines(pathToDataFile, StandardCharsets.UTF_8);
+
+        assertEquals(3, linesBeforeClear.size());
```
Подобные блоки ассертов можно, кстати, оборачивать в assertAll: будут выполнены все проверки в то время как без него выполниться только до первой упавшей 

https://askdev.ru/q/assertall-protiv-neskolkih-utverzhdeniy-v-junit5-290968/

EugeneLenkevich commented on src/ru/yandex/practicum/kanban/repository/impls/in_file/exceptions/InFileReadException.java:
```diff
@@ -0,0 +1,8 @@
+package ru.yandex.practicum.kanban.repository.impls.in_file.exceptions;
+
+public class InFileReadException extends RuntimeException {
```
Просто мысли в слух, а что означает тут приставка In, добавляет ли она какую то семантику, становится с ней код очевидней по сравнению если бы исключение называлось, например, FileReadException?

EugeneLenkevich reviewed (changes requested):
Привет, Александр! Спасибо за PR.

Работа впечатляет, все обстоятельно, правда надо ли было делать так сложно. В принципе , задумка авторов курса была в том чтобы отнаследовать предыдущий менеджер задач и в наследнике повесили сохранение и загрузку файлов.  То есть смыл именно в том, чтобы расширить существующую функциональность с помощью наследования. Но ладно, если получилось сделать вот так то и с более простым вариантом проблем не будет.

В любом случае все очень круто, только давай не будем обрабатывать как то исключения в тестах в `@Before`, пусть падает как есть, не надо переусложнять.

Ну и с Наступившим :)

AleksMosYanPrac committed:
refactor: work on remarks

rename user exceptions
wrap assertations in tests
delete exception handling from tests
rename manager to FileStoredTaskManager
add FileBackedTaskManager
add BackedRepository interface
add user "Manager exceptions"

AleksMosYanPrac committed:
style: work on checkstyle

AleksMosYanPrac committed:
style: work on checkstyle

EugeneLenkevich reviewed (approved):
Теперь все отлично, спасибо :)

AleksMosYanPrac commented:
> Теперь все отлично, спасибо :)

Спасибо!
